File: campus-navigate-app/flutter_app/lib/domain/tiled/map/png_map_model.dart
import 'package:indoor_navigation/domain/config/app_config.dart';
import 'package:indoor_navigation/domain/tiled/map/marker_model.dart';
import 'package:indoor_navigation/domain/tiled/map/object_overlay_model.dart';
import 'package:indoor_navigation/domain/tiled/map/path_segment_model.dart';
import 'package:indoor_navigation/domain/tiled/map/route_path_model.dart';
import 'package:indoor_navigation/domain/tiled/map/surface_model.dart';
import 'package:indoor_navigation/domain/tiled/map/tiled_map_layers.dart';
import 'package:indoor_navigation/domain/tiled/tiled_models.dart';

PngMapModel createPngMapModel(TiledMap map) {
  assertSupportedMap(map);

  final bounds = calculateChunkTileBounds(getVisibleTileLayers(map));
  final surface = createSurface(bounds, map);

  if (surface.width != demoPngSize.width ||
      surface.height != demoPngSize.height) {
    throw StateError(
      '${DemoPngSize.name} is ${demoPngSize.width}x${demoPngSize.height}, '
      'but the TMJ bounds are ${_formatDimension(surface.width)}x'
      '${_formatDimension(surface.height)}.',
    );
  }

  final routeNodes = getRouteNodes(map, surface);
  final blueMarker = createBlueMarkerState(routeNodes);
  final redMarker = createRedMarkerState(blueMarker);
  final routePath = createRoutePath(routeNodes);

  return PngMapModel(
    blueMarker: blueMarker,
    bounds: bounds,
    map: map,
    routePath: routePath,
    pathSegments: createPathSegmentsFromPoints(routePath),
    png: DemoPngMetadata(height: demoPngSize.height, width: demoPngSize.width),
    redMarker: redMarker,
    roomLabels: getRoomLabels(map, surface),
    routeNodes: routeNodes,
    surface: surface,
  );
}

PngMapModel createPngMapModelWithRoute(
  PngMapModel source,
  List<String> routeNodeIds,
) {
  final routePath = createRoutePath(source.routeNodes, nodeIds: routeNodeIds);
  final blueMarker = createBlueMarkerState(
    routePath,
    startNodeId: routeNodeIds.first,
  );
  return PngMapModel(
    blueMarker: blueMarker,
    bounds: source.bounds,
    map: source.map,
    routePath: routePath,
    pathSegments: createPathSegmentsFromPoints(routePath),
    png: source.png,
    redMarker: createRedMarkerState(blueMarker),
    roomLabels: source.roomLabels,
    routeNodes: source.routeNodes,
    surface: source.surface,
  );
}

String _formatDimension(double value) {
  return value == value.truncateToDouble()
      ? value.toInt().toString()
      : value.toString();
}

File: tarumt-nav-app/expo-app/src/tiled/model/pngMapModel.ts
import type { TiledMap } from '../type';
import { DEMO_PNG_SIZE } from './demoMapConfig';
import { createBlueMarkerState, createRedMarkerState } from './markerModel';
import { getRoomLabels, getRouteNodes } from './objectOverlayModel';
import { createPathSegmentsFromPoints } from './pathSegmentModel';
import { createRoutePath } from './routePathModel';
import { calculateChunkTileBounds, createSurface } from './surfaceModel';
import { assertSupportedMap, getVisibleTileLayers } from './tiledMapLayers';

export function createPngMapModel(map: TiledMap) {
  assertSupportedMap(map);

  const bounds = calculateChunkTileBounds(getVisibleTileLayers(map));
  const surface = createSurface(bounds, map);

  if (
    surface.width !== DEMO_PNG_SIZE.width ||
    surface.height !== DEMO_PNG_SIZE.height
  ) {
    throw new Error(
      `demo_1.png is ${DEMO_PNG_SIZE.width}x${DEMO_PNG_SIZE.height}, but the TMJ bounds are ${surface.width}x${surface.height}.`,
    );
  }

  const routeNodes = getRouteNodes(map, surface);
  const blueMarker = createBlueMarkerState(routeNodes);
  const redMarker = createRedMarkerState(blueMarker);
  const routePath = createRoutePath(routeNodes);

  return {
    blueMarker,
    bounds,
    map,
    routePath,
    pathSegments: createPathSegmentsFromPoints(routePath),
    png: DEMO_PNG_SIZE,
    redMarker,
    roomLabels: getRoomLabels(map, surface),
    routeNodes,
    surface,
  };
}
